### PR TITLE
Moved `array_merge` calls out of loops

### PR DESCRIPTION
**Pull request by simonberger**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no

`array_merge` calls in loops can potentially result in very slow performance and high memory usage

Here is a big loop example showing the possible big difference. The code I changed should not have such a big impact but it is a small improvement.

```php
<?php
$arr = [];

$startTime = microtime(true);
for ($i = 0; $i < 20000; ++$i) {
    $arr = array_merge($arr, ['a'.$i, 'b'.$i]);
}
$endTime = microtime(true);

echo "array_merge in loop took: " . number_format($endTime - $startTime, 3) . " Seconds \n";

$arr = [];

$startTime = microtime(true);
for ($i = 0; $i < 20000; ++$i) {
    $arr[] = ['a'.$i, 'b'.$i];
}
$arr = array_merge([], ...$arr);
$endTime = microtime(true);

echo "array_merge out of loop took: " . number_format($endTime - $startTime, 3) . "Seconds \n";
```

```
array_merge in loop took: 3.017 Seconds 
array_merge out of loop took: 0.008 Seconds 
```

### CODE REVIEW

**Comment by derrabus** [comment]
I believe we should not deliver this kind of optimization as a bugfix and target 3.2.x instead.

Did you benchmark your changes in any way? I agree that `array_merge()` in loops are a bad practise, but having some numbers could be good to justify the change.

**Comment by simonberger** [comment]
@derrabus 

> I believe we should not deliver this kind of optimization as a bugfix and target 3.2.x instead.

I don't mind rebasing it to 3.1.x but as it is a really small refactoring, most well covered by tests and easy to review it makes sense to me to already target 3.1.x.

> Did you benchmark your changes in any way? 

As you didn't mention the benchmark in the PR or even discredited it, I am not sure you noticed it?

**Comment by derrabus** [comment]
> > Did you benchmark your changes in any way?
> 
> As you didn't mention the benchmark in the PR

Oh sorry, I should've.

Your PR description contains a general benchmark for `array_merge()` in large loops. I acknowledge the positive impact of pulling `array_merge()` out of the loop in that scenario.

My question was if you had benchmaked _your changes_: Did you measure the impact of the changes contained in this PR on the DBAL's performance in typical scenarios? If not, that's completely fine. I was merely curious about your motivation. Maybe this change solved an actual bottleneck in your application.

Anyway, I'm fine with the changes. I'd still prefer targeting 3.2, but I think @morozov should decide that.

**Review by derrabus — DISMISSED** [review]

**Comment by simonberger** [comment]
@derrabus No I didn't benchmark it in scenarios I changed here. I did such a change in one of my projects after performance problems which improved it a lot.
If you use the PHP Inspections (EA Extended) Plugin of PHPStorm those array_merge calls are displayed as errors. If I notice a lot in projects that are important for me I fix them.

Let me know if I should use 3.2.x as the target.

File: src/Platforms/AbstractPlatform.php [hunk 1]
```diff
@@ -1768,19 +1768,18 @@ public function getCreateTableSQL(Table $table, $createFlags = self::CREATE_INDE
...
 
         foreach ($table->getColumns() as $column) {
-            if (
-                $this->_eventManager !== null
-                && $this->_eventManager->hasListeners(Events::onSchemaCreateTableColumn)
-            ) {
+            if ($hasSchemaCreateTableColumnListener) {
                 $eventArgs = new SchemaCreateTableColumnEventArgs($column, $table, $this);
 
                 $this->_eventManager->dispatchEvent(Events::onSchemaCreateTableColumn, $eventArgs);
```
Comment (simonberger) [inline src/Platforms/AbstractPlatform.php:1780]: Shall I add this phpstan error to the baseline or fix it with some inline phpdoc?
I thought phpstan is capable of noticing previous checks like this in `$hasSchemaCreateTableColumnListener`

**Comment by simonberger** [comment]
I added tests for some uncovered parts in AbstractPlatform. While doing this I noticed an unnecessary listener check done in the loop I removed an array_merge.
Because of those further refactoring I rebased to 3.2.x

**Comment by morozov** [comment]
> I'd still prefer targeting 3.2

This was my first thought when I saw the change. Absolutely agree.

File: src/Platforms/AbstractPlatform.php (hunk 1 above)
Comment (morozov) [reply src/Platforms/AbstractPlatform.php:1780]: Why does the event emitter care about whether there is a listener for its event or not? It should just emit the event. This way, this simple check could remain the the loop: `if ($this->_eventManager !== null)`.

Otherwise, I don't think that making this code more statically complicated is justified by a theoretical performance improvement.

File: src/Platforms/MySQLPlatform.php [hunk 2]
```diff
@@ -646,14 +646,15 @@ protected function getPreAlterTableIndexForeignKeySQL(TableDiff $diff)
...
-                $sql[] = $query;
+                $sql[] = [$query];
 
                 unset($diff->removedIndexes[$remKey], $diff->addedIndexes[$addKey]);
 
                 break;
             }
         }
 
+        $sql    = array_merge([], ...$sql);
```
Comment (morozov) [inline src/Platforms/MySQLPlatform.php:657]: What is this meant for?

File: src/Platforms/OraclePlatform.php [hunk 3]
```diff
@@ -407,28 +407,28 @@ protected function _getCreateTableSQL($name, array $columns, array $options = []
...
+            $sql[] = $this->getCreateAutoincrementSql($columnName, $name);
         }
 
         foreach ($indexes as $index) {
-            $sql[] = $this->getCreateIndexSQL($index, $name);
+            $sql[] = [$this->getCreateIndexSQL($index, $name)];
         }
 
-        return $sql;
+        return array_merge(parent::_getCreateTableSQL($name, $columns, $options), ...$sql);
```
Comment (morozov) [inline src/Platforms/OraclePlatform.php:431]: So now, instead of a `list<string>`, `$sql` becomes a `list<list<string>>`. I'm curious if the static analyzers can detect an attempt like `$sql[] = 'SELECT *'` as something invalid. If not, then this change will hurt maintainability more than it benefits performance.

File: src/Platforms/MySQLPlatform.php (hunk 2 above)
Comment (simonberger) [reply src/Platforms/MySQLPlatform.php:657]: This is how moving an `array_merge` out of the loop works. Instead of directly merging the arrays in the loop, an array of arrays is created which has to be unpacked to be put in an array_merge.
After this point `$sql` is the same it was before this patch.

The `[]` is necessary because otherwise there is an error if `$query` is an empty array.

**Comment by morozov** [comment]
I'm really on the fence about this change. I do appreciate the intent to improve the performance but it makes the code harder to understand and maintain. The benchmark in the description is synthetic: it calls `array_merge()` 20000 times in a loop without any IO. In real life, schema introspection always involves IO which is much slower than what's being optimized here.

For instance, the optimization being made in https://github.com/doctrine/dbal/pull/4882 will be much more substantial. I'm not saying "no", but I'm really curious to learn the real impact of this change on the performance.

File: src/Platforms/OraclePlatform.php (hunk 3 above)
Comment (simonberger) [reply src/Platforms/OraclePlatform.php:431]: We are in the internal scope of a function only.
Psalm/Phpstan blames if an unpacked array results in a string (like your example).
If you are afraid of further changes in those functions where someone does not look closely enough on the expectations of the variable and further help static analyzers we could do:

```php
/** @list<list<string>> */
$sql = [];
```

File: src/Platforms/AbstractPlatform.php (hunk 1 above)
Comment (derrabus) [reply src/Platforms/AbstractPlatform.php:1780]: Indeed, let's remove the `hasListeners()` call and move the null check back into the loop.

File: src/Platforms/OraclePlatform.php (hunk 3 above)
Comment (derrabus) [reply src/Platforms/OraclePlatform.php:431]: > I'm curious if the static analyzers can detect an attempt like `$sql[] = 'SELECT *'` as something invalid.

They don't. Both Psalm and PHPStan will detect and happily accept that the type of `$sql` will change from `list<list<string>>` to `list<list<string>|string>`. The `@var` annotation @simonberger suggests does not change that. However, both tools will render the final `array_merge()` call as invalid as a consequence.

tl;dr: The analyzers will complain that something is wrong, but they are not helpful in pinpointing the line that caused the issue.

File: src/Platforms/MySQLPlatform.php (hunk 2 above)
Comment (derrabus) [reply src/Platforms/MySQLPlatform.php:657]: But this `array_merge()` call is unnecessary, isn't it? It also means that the type of `$sql` changes from `list<list<string>>` to `list<string>` beyond this point. Sure, the static analyzers will catch that, but the human being @derrabus was a little bit confused at first.

File: src/Platforms/MySQLPlatform.php [hunk 4]
```diff
@@ -667,14 +668,12 @@ protected function getPreAlterTableIndexForeignKeySQL(TableDiff $diff)
             $diff->removedForeignKeys = [];
         }
 
-        $sql = array_merge(
+        return array_merge(
             $sql,
             $this->getPreAlterTableAlterIndexForeignKeySQL($diff),
             parent::getPreAlterTableIndexForeignKeySQL($diff),
             $this->getPreAlterTableRenameIndexForeignKeySQL($diff)
         );
```
Comment (derrabus) [inline src/Platforms/MySQLPlatform.php:676]: If we want to spare us the `array_merge()` call in line 657, we can do this:

```suggestion
        return array_merge(
            [],
            ...$sql,
            ...[
                $this->getPreAlterTableAlterIndexForeignKeySQL($diff),
                parent::getPreAlterTableIndexForeignKeySQL($diff),
                $this->getPreAlterTableRenameIndexForeignKeySQL($diff),
            ]
        );
```

File: src/Platforms/OraclePlatform.php [hunk 5]
```diff
@@ -407,28 +407,28 @@ protected function _getCreateTableSQL($name, array $columns, array $options = []
     {
         $indexes            = $options['indexes'] ?? [];
         $options['indexes'] = [];
-        $sql                = parent::_getCreateTableSQL($name, $columns, $options);
+        $sql                = [];
```
Comment (derrabus) [inline src/Platforms/OraclePlatform.php:410]: Let's leave the parent call up here.

```suggestion
        $sql                = [parent::_getCreateTableSQL($name, $columns, $options)];
```

File: src/Platforms/MySQLPlatform.php (hunk 2 above)
Comment (morozov) [reply src/Platforms/MySQLPlatform.php:657]: Now it's clear but it's totally unreadable. Sorry, but I think we should leave the code as is until it's proven that this patch provides any measurable performance benefits.
